src/box/lua/space.cc

```diff
@@ -616,26 +616,54 @@ box_lua_space_new(struct lua_State *L, struct space *space)
 		lua_getfield(L, -1, "space");
 	}
 	lua_rawgeti(L, -1, space_id(space));
-	if (lua_isnil(L, -1)) {
+	if (lua_isnil(L, -1) && space->lua_ref != LUA_NOREF) {
 		/*
-		 * If the space already exists, modify it, rather
-		 * than create a new one -- to not invalidate
-		 * Lua variable references to old space outside
-		 * the box.space[].
+		 * If the space does not exist and the space->lua_ref
+		 * contains a valid reference, that means we have a
+		 * space drop rollback. In this case we restore the
+		 * original lua space object.
 		 */
 		lua_pop(L, 1);
+		assert(tarantool_L == L);
```

Comment:
Nit: There's no need in this assertion (and below too). Any `L` is created from `tarantool_L` hence has the same globals.